pkg/pillar/hypervisor/xen.go

```diff
@@ -402,45 +402,14 @@ func (ctx xenContext) CreateDomConfig(domainName string, config types.DomainConf
 }
 
 func (ctx xenContext) Stop(domainName string, domainID int, force bool) error {
-	log.Infof("xlShutdown %s %d\n", domainName, domainID)
-	args := []string{
-		"xl",
-		"shutdown",
-		domainName,
-	}
-	if force {
-		args = append(args, "-F")
-	}
-	ctrdCtx, done := ctx.ctrdClient.CtrNewUserServicesCtx()
-	defer done()
-	stdOut, stdErr, err := ctx.ctrdClient.CtrExec(ctrdCtx, domainName, args)
-	if err != nil {
-		log.Errorln("xl shutdown failed ", err)
-		log.Errorln("xl shutdown output ", stdOut, stdErr)
-		return fmt.Errorf("xl shutdown failed: %s %s", stdOut, stdErr)
-	}
-	log.Infof("xl shutdown done\n")
-	return nil
+	return ctx.ctrdContext.Stop(domainName, domainID, true)
 }
 
 func (ctx xenContext) Delete(domainName string, domainID int) error {
-	log.Infof("xlDestroy %s %d\n", domainName, domainID)
-	ctrdSystemCtx, done := ctx.ctrdClient.CtrNewSystemServicesCtx()
-	defer done()
-	stdOut, stdErr, err := ctx.ctrdClient.CtrSystemExec(ctrdSystemCtx, "xen-tools",
-		[]string{"xl", "destroy", domainName})
-	if err != nil {
-		log.Errorln("xl destroy failed ", err)
-		log.Errorln("xl destroy output ", stdOut, stdErr)
-		return fmt.Errorf("xl destroy failed: %s %s", stdOut, stdErr)
-	}
-	log.Infof("xl destroy done %s %d\n", domainName, domainID)
-
 	// now lets take care of the task itself
 	if err := ctx.ctrdContext.Stop(domainName, domainID, true); err != nil {
```

Comment:
Here force=true makes sense, but I still don't understand why for KVM we need additional code and not here.

Reply:
We do need additional code here as well. At least with how everything else is laid out. @zed-rishabh is jut relying on the fact that when resident xl gets killed violently -- it *seems* to also kill domain. I've observed that too, but I really don't think it is documented behaviour.